staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -1035,18 +1029,24 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
 //  * http.Client.Do errors are returned directly.
 func (r *Request) Do(ctx context.Context) Result {
 	var result Result
+	// Metrics for total request latency
+	start := time.Now()
 	err := r.request(ctx, func(req *http.Request, resp *http.Response) {
 		result = r.transformResponse(resp, req)
 	})
 	if err != nil {
 		return Result{err: err}
 	}
+	metrics.RequestLatency.Observe(ctx, r.verb, r.finalURLTemplate(), time.Since(start))
+	metrics.RequestSize.Observe(ctx, r.verb, r.finalURLTemplate(), float64(len(result.body)))
```

Comment:
can we put it in a `defer` block to maintain current behavior?

Reply:
current behavior is wrong, you can't measure latency if there is no response or if you didn't connect to the server, per example, if the request timed out because a network error, we were recording its latency

Reply:
> current behavior is wrong, you can't measure latency if there is no response or if you didn't connect to the server, per example, if the request timed out because a network error, we were recording its latency

I can buy that.  Do I get a new metric measuring timeouts?

Reply:
we already have it
https://github.com/kubernetes/kubernetes/blob/15112c7286497cc42b5f0552b8dc5c935c6cac04/staging/src/k8s.io/client-go/rest/request.go#L791-L795

Reply:
current behavior includes error (`err` is set) where the network peer has been reached and something went wrong while the request was being processed. i think we should include these in the metrics

https://github.com/kubernetes/kubernetes/blob/9918aa1e035a00bc7c0f16a05e1b222650b3eabc/staging/src/k8s.io/client-go/rest/request.go#L1024-L1026

Reply:
I think that err comes from

https://github.com/kubernetes/kubernetes/blob/9918aa1e035a00bc7c0f16a05e1b222650b3eabc/staging/src/k8s.io/client-go/rest/request.go#L980-L986

and the docs say https://pkg.go.dev/net/http#Client.Do
```
func (*Client) Do ¶
func (c *Client) Do(req *Request) (*Response, error)
Do sends an HTTP request and returns an HTTP response, following policy (such as redirects, cookies, auth) as configured on the client.

An error is returned if caused by client policy (such as CheckRedirect), or failure to speak HTTP (such as a network connectivity problem). A non-2xx status code doesn't cause an error.
```

Reply:
> An error is returned if caused by client policy (such as CheckRedirect), or failure to speak HTTP (such as a network connectivity problem)

err is not something we want to use for measuring latency

Reply:
nevermind, I kept the same bahevaviour for the latency

Reply:
@tkashem you were right I was wrong, it seems the common practices is to track latency also for failing request , I removed the commit changing that